pylint/checkers/base_checker.py

```diff
@@ -0,0 +1,114 @@
+# Copyright (c) 2006-2014 LOGILAB S.A. (Paris, FRANCE) <contact@logilab.fr>
+# Copyright (c) 2013-2014 Google, Inc.
+# Copyright (c) 2013 buck@yelp.com <buck@yelp.com>
+# Copyright (c) 2014-2017 Claudiu Popa <pcmanticore@gmail.com>
+# Copyright (c) 2014 Brett Cannon <brett@python.org>
+# Copyright (c) 2014 Arun Persaud <arun@nubati.net>
+# Copyright (c) 2015 Ionel Cristian Maries <contact@ionelmc.ro>
+# Copyright (c) 2016 Moises Lopez <moylop260@vauxoo.com>
+# Copyright (c) 2017-2018 Bryce Guinta <bryce.paul.guinta@gmail.com>
+
+# Licensed under the GPL: https://www.gnu.org/licenses/old-licenses/gpl-2.0.html
+# For details: https://github.com/PyCQA/pylint/blob/master/COPYING
+
+from typing import Any
+
+from pylint.config import OptionsProviderMixIn
+from pylint.exceptions import InvalidMessageError
+from pylint.interfaces import UNDEFINED
+from pylint.message import build_message_definition
+
+
+class BaseChecker(OptionsProviderMixIn):
+
+    # checker name (you may reuse an existing one)
+    name = None  # type: str
+    # options level (0 will be displaying in --help, 1 in --long-help)
+    level = 1
+    # ordered list of options to control the ckecker behaviour
+    options = ()  # type: Any
+    # messages issued by this checker
+    msgs = {}  # type: Any
+    # reports issued by this checker
+    reports = ()  # type: Any
+    # mark this checker as enabled or not.
+    enabled = True
+
+    def __init__(self, linter=None):
+        """checker instances should have the linter as argument
+
+        :param ILinter linter: is an object implementing ILinter."""
+        if self.name is not None:
+            self.name = self.name.lower()
+        OptionsProviderMixIn.__init__(self)
+        self.linter = linter
+
+    def __gt__(self, other):
```

Comment:
Right. I added two commits related to moving build_message_definition, when I should not have because I abandoned that part, so this make the PR hard to understand. 

* build_message_definition is used in `pyliny.checker.base_checker` and also in `message.MessageHandlerMixIn.print_full_documentation`.
* The idea is to put it in the checker packages only to break the cycle
* So we need to use a Checker instance in print_full_documentation
* There is a call to sorted() on a list of checker name here : https://github.com/PyCQA/pylint/blob/master/pylint/message/message_handler_mix_in.py#L399
* So we need to implement `__gt__` in `BaseChecker`
* Doing the refactoring to take checker instead of strings representing the checker name makes MyPy realize that MessageHandlerMixIn do not have a get_checkers() function so it would need to be refactored again to not break the tests suite (As explained here : https://github.com/PyCQA/pylint/pull/2809#issuecomment-475983034)
* So I reverted the changes from the point we took a checker instances instead of a string (Branch with said change is here : https://github.com/Pierre-Sassoulas/pylint/tree/build-message-definition-in-checker)

I did not revert the two commits that were working but are part of the change for build_message_definition (https://github.com/PyCQA/pylint/pull/2809/commits/45352e37ae5c533929a344fd6523f5ac84428fc6 and https://github.com/PyCQA/pylint/pull/2809/commits/be9f0eb53f7504a557d644fb1f49719c1d9ccac7).

Let me know if you want me to revert them.

Reply:
Thanks for the extra context, makes much more sense now.